### ISSUE TITLE
Switch to TOML for configuration file

### ISSUE THREAD

**Issue by chrisduerr** [alacritty/alacritty#6592]
I'm quite honestly tired of looking at YAML files with broken indentation. While YAML is an awful format even for experienced users, clearly it's completely impossible for inexperienced ones to grasp its concept. And the documentation of our configuration file with everything commented out by default probably doesn't help.

Since as far as I know all previous issues with TOML have been fixed and escape sequences can be easily specified in TOML strings now, I think it would make sense to switch to TOML as our default format.

In a transitory period we could probably get away with just loading existing YAML files and writing them out as TOML, to allow for backward compatibility. This will probably produce somewhat messy TOML files, but I can't imagine an auto-generated TOML file would be much worse than a YAML file.

The only major drawbacks I see is that this is a rather big change that will no doubt annoy a bunch of people and some more advanced features of YAML (like anchors) are not supported in TOML. But nobody *really* needs anchors and I kinda like annoying people.

@kchibisov You proposed this a while ago on IRC, do you think this is a bad idea?

See https://github.com/alacritty/alacritty/issues/1217 for prior discussions.

**Comment by kchibisov** [comment]
>@kchibisov You proposed this a while ago on IRC, do you think this is a bad idea?

I do like toml more. In general what we can do is to have support for both at the same time with a warning to do a conversion with some alacritty flag. But having both at the same time doesn't sound like fun to maintain.

I do wonder how stuff like multiple bindings will end up working.

>Since as far as I know all previous issues with TOML have been fixed and escape sequences can be easily specified in TOML strings now, I think it would make sense to switch to TOML as our default format.

Is that supported by serde at least? Though, patching shouldn't be a problem.

**Comment by keltia** [comment]
[HCL](https://developer.hashicorp.com/terraform/language) is also a good if not better choice.

**Comment by skyfaller** [comment]
Aw man, I use anchors heavily in my Alacritty yaml config, for changing themes with scripting. I'm not quite sure what the best way to do this with TOML is. Do you have any suggestions for rewriting something like this in TOML?

```yaml
schemes:
  # Colors (Blood Moon)
  blood_moon: &dark
    # Default colors:
    primary:
      background: '#10100E'
      foreground: '#C6C6C4'

    # Normal colors:
    normal:
      black:   '#10100E'
      red:     '#C40233'
      green:   '#009F6B'
      yellow:  '#FFD700'
      blue:    '#0087BD'
      magenta: '#9A4EAE'
      cyan:    '#20B2AA'
      white:   '#C6C6C4'

    # Bright colors:
    bright:
      black:   '#696969'
      red:     '#FF2400'
      green:   '#03C03C'
      yellow:  '#FDFF00'
      blue:    '#007FFF'
      magenta: '#FF1493'
      cyan:    '#00CCCC'
      white:   '#FFFAFA'

  # Colors (Terminal.app Basic)
  terminal_app_basic: &light
    primary:
      background: '#FFFFFF'
      foreground: '#000000'
    normal:
      black:      '#000000'
      red:        '#990000'
      green:      '#00A600'
      yellow:     '#999900'
      blue:       '#0000B2'
      magenta:    '#B200B2'
      cyan:       '#00A6B2'
      white:      '#BFBFBF'
    bright:
      black:      '#666666'
      red:        '#E50000'
      green:      '#00D900'
      yellow:     '#D0D000'
      blue:       '#0000FF'
      magenta:    '#E500E5'
      cyan:       '#00E5E5'
      white:      '#E5E5E5'

  # Colors (Oceanic Next)
  oceanic_next: &blue
    # Default colors:
    primary:
      background: '#1b2b34'
      foreground: '#d8dee9'

    # Colors the cursor will use if `custom_cursor_colors: ` is true
    cursor:
      text: '#1b2b34'
      cursor: '#ffffff'

    # Normal colors:
    normal:
      black:   '#343d46'
      red:     '#EC5f67'
      green:   '#99C794'
      yellow:  '#FAC863'
      blue:    '#6699cc'
      magenta: '#c594c5'
      cyan:    '#5fb3b3'
      white:   '#d8dee9'

    # Bright colors:
    bright:
      black:   '#343d46'
      red:     '#EC5f67'
      green:   '#99C794'
      yellow:  '#FAC863'
      blue:    '#6699cc'
      magenta: '#c594c5'
      cyan:    '#5fb3b3'
      white:   '#d8dee9'

# Colors (Gruvbox light)
  gruvbox_light: &yellow
    # Default colors:
    primary:
      # hard contrast: background = '#f9f5d7'
      background: '#fbf1c7'
      # soft contrast: background = '#f2e5bc'
      foreground: '#3c3836'

    # Normal colors:
    normal:
      black:   '#fbf1c7'
      red:     '#cc241d'
      green:   '#98971a'
      yellow:  '#d79921'
      blue:    '#458588'
      magenta: '#b16286'
      cyan:    '#689d6a'
      white:   '#7c6f64'

    # Bright colors:
    bright:
      black:   '#928374'
      red:     '#9d0006'
      green:   '#79740e'
      yellow:  '#b57614'
      blue:    '#076678'
      magenta: '#8f3f71'
      cyan:    '#427b58'
      white:   '#3c3836'

  # Colors (Tango)
  tango: &green
    primary:
      background: '#000000'
      foreground: '#00ff00'

    # Normal colors:
    normal:
        black:   '#2e3436'
        red:     '#cc0000'
        green:   '#73d216'
        yellow:  '#edd400'
        blue:    '#3465a4'
        magenta: '#75507b'
        cyan:    '#06989a'
        white:   '#d3d7cf'

    # Bright colors:
    bright:
        black:   '#2e3436'
        red:     '#ef2929'
        green:   '#8ae234'
        yellow:  '#fce94f'
        blue:    '#729fcf'
        magenta: '#ad7fa8'
        cyan:    '#34e2e2'
        white:   '#eeeeec'

  # Colors (Monokai Soda)
  monokai_soda: &purple
    # Default colors:
    primary:
      background: '#1a1a1a'
      foreground: '#c4c5b5'

    # Normal colors:
    normal:
      black:   '#1a1a1a'
      red:     '#f4005f'
      green:   '#98e024'
      yellow:  '#fa8419'
      blue:    '#9d65ff'
      magenta: '#f4005f'
      cyan:    '#58d1eb'
      white:   '#c4c5b5'

    # Bright colors:
    bright:
      black:   '#625e4c'
      red:     '#f4005f'
      green:   '#98e024'
      yellow:  '#e0d561'
      blue:    '#9d65ff'
      magenta: '#f4005f'
      cyan:    '#58d1eb'
      white:   '#f6f6ef'

  # Colors (Darkside)
  darkside: &red
    primary:
      background: '#222324'
      foreground: '#BABABA'

    # Normal colors
    normal:
      black:    '#000000'
      red:      '#E8341C'
      green:    '#68C256'
      yellow:   '#F2D42C'
      blue:     '#1C98E8'
      magenta:  '#8E69C9'
      cyan:     '#1C98E8'
      white:    '#BABABA'

    # Bright colors
    bright:
      black:    '#666666'
      red:      '#E05A4F'
      green:    '#77B869'
      yellow:   '#EFD64B'
      blue:     '#387CD3'
      magenta:  '#957BBE'
      cyan:     '#3D97E2'
      white:    '#BABABA'

  dracula: &lavender
    primary:
      background: '0x282a36'
      foreground: '0xf8f8f2'
    cursor:
      text: CellBackground
      cursor: CellForeground
    vi_mode_cursor:
      text: CellBackground
      cursor: CellForeground
    search:
      matches:
        foreground: '0x44475a'
        background: '0x50fa7b'
      focused_match:
        foreground: '0x44475a'
        background: '0xffb86c'
      bar:
        background: '0x282a36'
        foreground: '0xf8f8f2'
    line_indicator:
      foreground: None
      background: None
    selection:
      text: CellForeground
      background: '0x44475a'
    normal:
      black:   '0x000000'
      red:     '0xff5555'
      green:   '0x50fa7b'
      yellow:  '0xf1fa8c'
      blue:    '0xbd93f9'
      magenta: '0xff79c6'
      cyan:    '0x8be9fd'
      white:   '0xbfbfbf'
    bright:
      black:   '0x4d4d4d'
      red:     '0xff6e67'
      green:   '0x5af78e'
      yellow:  '0xf4f99d'
      blue:    '0xcaa9fa'
      magenta: '0xff92d0'
      cyan:    '0x9aedfe'
      white:   '0xe6e6e6'
    dim:
      black:   '0x14151b'
      red:     '0xff2222'
      green:   '0x1ef956'
      yellow:  '0xebf85b'
      blue:    '0x4d5b86'
      magenta: '0xff46b0'
      cyan:    '0x59dffc'
      white:   '0xe6e6d1'

# Dark options: dark,green,purple,blue,red,lavender
# Light options: light,yellow
colors: *light
# To change color theme using sd, try:
# sd '^colors: *.*$' 'colors: *light' ~/.config/alacritty/alacritty.yml
# I've also defined a fish function "theme-alacritty"
```

**Comment by kchibisov** [comment]
@skyfaller use imports.

**Comment by skyfaller** [comment]
Thank you for the tip. For anyone who needs more detail on imports, some documentation: https://github.com/alacritty/alacritty/blob/master/extra/man/alacritty.5.scd#general

Any chance you could show some examples of how imports are used? Is there a sample config file somewhere?

**Comment by chrisduerr** [comment]
If you're looking for toml syntax docs, you can find array docs here: https://toml.io/en/v1.0.0#array

Other than that it's just exactly as documented in the manpage.

**Comment by pi-rho** [comment]
Is it hard to support both? TOML pokes me in the eye while YAML is super comfy. Everyone has their preferences. Can't serde shield Alacritty and allow users to pick their favorite?

**Comment by kchibisov** [comment]
@pi-rho you better not know how many issues we've got because `option is not working` and `broken indention with yml`, it's not fun to spend our time on trouble shooting them, with toml we warn on unused keys and you don't have to deal with indention, and you really don't have to deal with configuring alacritty that much anyway. It's true that generated config might be a bit off, but hand written ones are not that bad.

**Comment by pi-rho** [comment]
That may be so, but this is where I'm at today. ;)

```patch
diff --git i/alacritty/src/config/mod.rs w/alacritty/src/config/mod.rs
index 821e9b6b..312b8828 100644
--- i/alacritty/src/config/mod.rs
+++ w/alacritty/src/config/mod.rs
@@ -228,7 +228,7 @@ pub fn deserialize_config(path: &Path) -> Result<Value> {
     // Convert YAML to TOML as a transitionary fallback mechanism.
     let extension = path.extension().unwrap_or_default();
     if (extension == "yaml" || extension == "yml") && !contents.trim().is_empty() {
-        warn!("YAML config {path:?} is deprecated, please migrate to TOML");
+        //warn!("YAML config {path:?} is deprecated, please migrate to TOML");

         let value: serde_yaml::Value = serde_yaml::from_str(&contents)?;
         contents = toml::to_string(&value)?;
```

**Comment by chrisduerr** [comment]
Sooner or later support for YAML will be dropped. There is no reason why we'd maintain both and even the current YAML implementation is just a subpar fill-in for the transition period.

**Comment by nickeb96** [comment]
Do I need to do something specific to enable the TOML escape sequences mentioned above?  I'm getting an error when I try to use them:

```text
[8215.690349958s] [ERROR] [alacritty_config_derive] Unable to load config
Config error: TOML parse error at line 122, column 12
    |
122 | chars = "\e[13;2u"
    |            ^
invalid escape sequence
expected `b`, `f`, `n`, `r`, `t`, `u`, `U`, `\`, `"`
```

**Comment by chrisduerr** [comment]
`\e` is not supported in TOML (yet). You need to use `\u001b` instead.

Both `\e` and `\x1b` will be supported with the next TOML release.

**Comment by bluz71** [comment]
> And the documentation of our configuration file with everything commented out by default probably doesn't help.

Can an `alacritty.toml` be created in this base of this repository mimicking the previous `alacritty.yml` file?  That is with all options listed and mostly commented out?

I can't tell how many times I referred to that file to tweak a setting here and there.

That document was *the* document I referred to as the configuration source of truth. Now I feel lost.

**Comment by kchibisov** [comment]
>That document was the document I referred to as the configuration source of truth. Now I feel lost.

But you have man page now with **all** the options? And a separate for all default bindings.

**Comment by bluz71** [comment]
Ok, I had a look now. The `alacritty.5` man page has the information, so that is good to know it exists somewhere.

I don't think it is quite is quick and easy as the previous `alacritty.yml` example/instance file. But opinions will vary.

Thanks for the knowledge.

**Comment by kchibisov** [comment]
But you can type `man 5 alacritty` in your terminal and have all the docs without a need to go on a github.com, because config before wasn't shipped even.

**Comment by Lucksinia** [comment]
...I am sorry for opening this one, but `No manual entry for alacritty in section 5`?

> But you can type `man 5 alacritty` in your terminal and have all the docs without a need to go on a github.com, because config before wasn't shipped even.

**Comment by kchibisov** [comment]
>I am sorry for opening this one, but No manual entry for alacritty in section 5?

if you've built from source build man pages as well, no stable release was made with toml yet.

**Comment by pi-rho** [comment]
my eyes are bleeding

```toml
[[keyboard.bindings]]
action = "ClearSelection"
key = "C"
mode = "Vi|~Search"
mods = "Control|Shift"

[[keyboard.bindings]]
action = "PasteSelection"
key = "Insert"
mods = "Shift"

[[keyboard.bindings]]
action = "ResetFontSize"
key = "Key0"
mods = "Control"

[[keyboard.bindings]]
action = "IncreaseFontSize"
key = "Equals"
mods = "Control"
```

:)

**Comment by kchibisov** [comment]
@pi-rho 

you're free to format the config however you want, you can use something like that

```toml
[keyboard]
bindings = [
  { key = "C", mods = "Control|Shift", mode = "Vi|Search", action = "ClearSelection" },
  { key = "Insert", mods = "Shift", action = "Paste" }
]
```

**Comment by cy8aer** [comment]
It would be nice to have an example toml now because you are not able to prepare a color scheme based theming without the knowledge how to transform it from yaml. Migration makes the new toml unfunctional for example.

**Comment by chrisduerr** [comment]
> Migration makes the new toml unfunctional for example.

Please open an issue if you've found a bug in the `alacritty migrate` command.

If you're talking about links no longer working, then that is an intended feature and there is not going to be any replacement for this in toml. You should be using imports instead (even with yaml really).

**Comment by cy8aer** [comment]
ok, I thought it should be a replacement and therefor please close the issue (and I for myself stay at yaml). Thanks

**Comment by andrewriveradev** [comment]
@cy8aer 

```toml

[keyboard]
bindings = [
    { key = "Paste", action = "Paste" },
    { key = "Copy", action = "Copy" },
    { key = "L", mods = "Control", action = "ClearLogNotice" },
    { key = "L", mods = "Control", mode = "~Vi|~Search", chars = "\\u000C" },
    { key = "PageUp", mods = "Shift", mode = "~Alt", action = "ScrollPageUp" },
    { key = "PageDown", mods = "Shift", mode = "~Alt", action = "ScrollPageDown" },
    { key = "Home", mods = "Shift", mode = "~Alt", action = "ScrollToTop" },
    { key = "End", mods = "Shift", mode = "~Alt", action = "ScrollToBottom" },
    { key = "Space", mods = "Shift|Control", mode = "~Search", action = "ToggleViMode" },
    { key = "Space", mods = "Shift|Control", mode = "Vi|~Search", action = "ScrollToBottom" },
    { key = "Escape", mode = "Vi|~Search", action = "ClearSelection" },
    { key = "I", mode = "Vi|~Search", action = "ToggleViMode" },
    { key = "I", mode = "Vi|~Search", action = "ScrollToBottom" },
    { key = "C", mods = "Control", mode = "Vi|~Search", action = "ToggleViMode" },
    { key = "Y", mods = "Control", mode = "Vi|~Search", action = "ScrollLineUp" },
    { key = "E", mods = "Control", mode = "Vi|~Search", action = "ScrollLineDown" },
    { key = "G", mode = "Vi|~Search", action = "ScrollToTop" },
    { key = "G", mods = "Shift", mode = "Vi|~Search", action = "ScrollToBottom" },
    { key = "B", mods = "Control", mode = "Vi|~Search", action = "ScrollPageUp" },
    { key = "F", mods = "Control", mode = "Vi|~Search", action = "ScrollPageDown" },
    { key = "U", mods = "Control", mode = "Vi|~Search", action = "ScrollHalfPageUp" },
    { key = "D", mods = "Control", mode = "Vi|~Search", action = "ScrollHalfPageDown" },
    { key = "Y", mode = "Vi|~Search", action = "Copy" },
    { key = "Y", mode = "Vi|~Search", action = "ClearSelection" },
    { key = "Copy", mode = "Vi|~Search", action = "ClearSelection" },
    { key = "V", mode = "Vi|~Search", action = "ToggleNormalSelection" },
    { key = "V", mods = "Shift", mode = "Vi|~Search", action = "ToggleLineSelection" },
    { key = "V", mods = "Control", mode = "Vi|~Search", action = "ToggleBlockSelection" },
    { key = "V", mods = "Alt", mode = "Vi|~Search", action = "ToggleSemanticSelection" },
    { key = "Return", mode = "Vi|~Search", action = "Open" },
    { key = "Z", mode = "Vi|~Search", action = "CenterAroundViCursor" },
    { key = "K", mode = "Vi|~Search", action = "Up" },
    { key = "J", mode = "Vi|~Search", action = "Down" },
    { key = "H", mode = "Vi|~Search", action = "Left" },
    { key = "L", mode = "Vi|~Search", action = "Right" },
    { key = "Up", mode = "Vi|~Search", action = "Up" },
    { key = "Down", mode = "Vi|~Search", action = "Down" },
    { key = "Left", mode = "Vi|~Search", action = "Left" },
    { key = "Right", mode = "Vi|~Search", action = "Right" },
    { key = "Key0", mode = "Vi|~Search", action = "First" },
    { key = "Key4", mods = "Shift", mode = "Vi|~Search", action = "Last" },
    { key = "Key6", mods = "Shift", mode = "Vi|~Search", action = "FirstOccupied" },
    { key = "H", mods = "Shift", mode = "Vi|~Search", action = "High" },
    { key = "M", mods = "Shift", mode = "Vi|~Search", action = "Middle" },
    { key = "L", mods = "Shift", mode = "Vi|~Search", action = "Low" },
    { key = "B", mode = "Vi|~Search", action = "SemanticLeft" },
    { key = "W", mode = "Vi|~Search", action = "SemanticRight" },
    { key = "E", mode = "Vi|~Search", action = "SemanticRightEnd" },
    { key = "B", mods = "Shift", mode = "Vi|~Search", action = "WordLeft" },
    { key = "W", mods = "Shift", mode = "Vi|~Search", action = "WordRight" },
    { key = "E", mods = "Shift", mode = "Vi|~Search", action = "WordRightEnd" },
    { key = "Key5", mods = "Shift", mode = "Vi|~Search", action = "Bracket" },
    { key = "Slash", mode = "Vi|~Search", action = "SearchForward" },
    { key = "Slash", mods = "Shift", mode = "Vi|~Search", action = "SearchBackward" },
    { key = "N", mode = "Vi|~Search", action = "SearchNext" },
    { key = "N", mods = "Shift", mode = "Vi|~Search", action = "SearchPrevious" },
]

```

**Comment by papricasix** [comment]
I'm quite unhappy about the change to `toml`. It does not solve any real problems and forces everyone to change their config. I don't prefer `yaml`, but it's just a waste of time for every side (change of the config as well as programming a solution to a non-problem). :disappointed: 
Nevertheless I really like Alacritty, and want to thank everyone involved in it's development.

**Comment by chrisduerr** [comment]
> It does not solve any real problems

Clearly you're not following the Alacritty issue tracker.

**Comment by chenasraf** [comment]
Thanks for taking a perfectly valid format and replacing it with crap

**Comment by bluz71** [comment]
> Thanks for taking a perfectly valid format and replacing it with crap

Not perfect if you had an extra space somewhere in one of the indents.

**Comment by chenasraf** [comment]
> > Thanks for taking a perfectly valid format and replacing it with crap
> 
> Not perfect if you had an extra space somewhere in one of the indents.

Prettier and .editorconfig (+many more alternatives) are a thing.

Using this format without proper syntax highlights for the unicode strings (which are also less readable and too similar to each other to properly differentiate), however, is just not convenient at all to work with when you're re-doing more than 1 key mapping change at any given time.

Also, the output from the migration command isn't the same as the one being explained in the docs, so that's even more confusing, unless you go back and manually update (possibly dozens) of shortcuts, each now taking multiple lines and not one, which makes it harder to update in batches, unless that's how you like spending your day

**Comment by chrisduerr** [comment]
> Using this format which doesn't have proper syntax highlights for the unicode strings

Syntax highlighting is 100% up to you and your editor. TOML has perfectly good syntax highlighting in any editor I've ever seen. It's not some obscure unknown format, it's extremely widespread.

> Also, the output from the migration command isn't the same as the one being explained in the docs

It shouldn't come as a surprise that a computer-automated migration will not produce a perfect config for every human out there. It's a point to get you going and start out from, so everything works and it's up to you to format it however you prefer.

**Comment by drbachler** [comment]
I am so sick and tired of things always changing and breaking on me.
For what? "... and I kinda like annoying people". Your mission is a success.

**Comment by chrisduerr** [comment]
> I am so sick and tired of things always changing and breaking on me.

Then don't use software in beta? Is that such a difficult concept to grasp?

**Comment by ManuGithubSteam** [comment]
slient-rock:~$alacritty migrate
Configuration file migration failed:
    "/home/manuel/.alacritty.yml": parsing error: Yaml conversion error: unsupported unit type

What does this mean ? It does not say wich line is wrong ?
Any help appreciated :-)
Thanks!

**Comment by kchibisov** [comment]
@ManuGithubSteam the unit type issue means that you have something with a `null` in your yaml, an example would be

```yaml
window:
  cursor:
  opacity: 0.9
```

`cursor:` is a unit type here, there's a https://github.com/alacritty/alacritty/pull/7536 to just drop them.

**Comment by ManuGithubSteam** [comment]
> @ManuGithubSteam the unit type issue means that you have something with a `null` in your yaml, an example would be
> 
> ```yaml
> window:
>   cursor:
>   opacity: 0.9
> ```
> 
> `cursor:` is a unit type here, there's a #7536 to just drop them.

I rewrote my config in a dich effort because i need it for work tomorrow. Thankfully someone posted an example. If the project aims for wider adoption something like a example  conf should be common sense.

Thanks for your help tho!

**Comment by chrisduerr** [comment]
Not having your original config completely broken is usually a good start for things to work smoothly.

**Comment by J-MR-T** [comment]
> > I am so sick and tired of things always changing and breaking on me.
> 
> Then don't use software in beta? Is that such a difficult concept to grasp?

The readme says

> The software is considered to be at a beta level of readiness; *there are a few missing features and bugs to be fixed*

not

> The software is considered to be at a beta level of readiness; we "kinda like annoying people" so be prepared to be annoyed.

---

To be clear, I don't even question the migration itself, breaking changes are sometimes beneficial or even required; personally, I also don't mind migrating some alacritty config files once per year, but this comment really threw me off. 

TL;DR: I don't like using software that wants to annoy me.

**Comment by bluz71** [comment]
I found the migration simple and effective, a small amount of customization later, 10 mins all in all. Easy-peasy.

I suspect there is a silent majority where that is the case.

I am pleased with the change, TOML is a far better config format in comparison to YAML.

**Comment by J-MR-T** [comment]
As I said, I don't question this migration itself. I question the attitude of liking to annoy people. 

Side note, I'm sure you could read the comment about annoying people as a joke, but even if it was, that sort of joke only works, if there is a kernel of truth to it.

**Comment by aarav2you** [comment]
> [HCL](https://developer.hashicorp.com/terraform/language) is also a good if not better choice.

Unfortunately your opinion is flat out wrong

**Comment by zeionara** [comment]
I hate the TOML format. How to disable the fcking warning?

**Comment by zeionara** [comment]
> That may be so, but this is where I'm at today. ;)
> 
> ```diff
> diff --git i/alacritty/src/config/mod.rs w/alacritty/src/config/mod.rs
> index 821e9b6b..312b8828 100644
> --- i/alacritty/src/config/mod.rs
> +++ w/alacritty/src/config/mod.rs
> @@ -228,7 +228,7 @@ pub fn deserialize_config(path: &Path) -> Result<Value> {
>      // Convert YAML to TOML as a transitionary fallback mechanism.
>      let extension = path.extension().unwrap_or_default();
>      if (extension == "yaml" || extension == "yml") && !contents.trim().is_empty() {
> -        warn!("YAML config {path:?} is deprecated, please migrate to TOML");
> +        //warn!("YAML config {path:?} is deprecated, please migrate to TOML");
> 
>          let value: serde_yaml::Value = serde_yaml::from_str(&contents)?;
>          contents = toml::to_string(&value)?;
> ```

Thank you! I despise the TOML format. It is pain to read and edit it. No other option than to disable the annoying warning in source.

**Comment by ikouchiha47** [comment]
Which is fine. migrating to toml. But I hope I find a migration tool from yaml to toml. I mean I have had my dotfiles there for nearly half a decade now. Only change was for neovim. But then it was an eniterly different thing. Who is going to change the config manually now. Its easier to just find another terminal. :laughing:

**Comment by chrisduerr** [comment]
If only Alacritty devs had integrated an automatic migration tool specifically built to ease this transition for existing users.

**Comment by anonhostpi** [comment]
I don't understand the practice of not supporting both config formats. Even moreso, _deprecating_ a defacto standard.

Even if you are running into the problem of "too many indentation issues," then write a better SEO-optimized wiki, or add the problem to the issue template. It shouldn't take a lot of effort to educate end users on the _defacto standard_ for configuration files.

Hell, you probably don't need to do more than update the issue template, because there are plenty of free education tools for YAML. **_Even linking a simple YAML validator in the issue template like https://www.yamllint.com/ should fix the original problem statement._**

I will say that adding and switching the _default_ configuration language to toml makes sense, because of the aforementioned better support for character escaping, but fully deprecating yaml does not.

You can't really justify deprecation with "can't support" either, because yaml and toml are just wrappers for json. If you can support one of the 3 (yaml, toml, or json), your app can support the other 2.

You might justify it with "don't want to support," but even then that is resolved with the above statement "write a better SEO-optimized wiki, or add the problem to the issue template"

**Comment by kchibisov** [comment]
>Even if you are running into the problem of "too many indentation issues," then write a better SEO-optimized wiki, or add the problem to the issue template. It shouldn't take a lot of effort to educate end users on the defacto standard for configuration files.

No one ever reads, remember that, people who read don't have issues with any of that. Alacritty is also not a stable software, as in pre 1.0, so we don't have to bother with legacy, this is stated in README and it was like that for a long time. After 1.0 no one will ever change the config format.

>I will say that adding and switching the default configuration language to toml does make sense, because of the aforementioned better support for character escaping, but fully deprecating yaml does not.

toml is much simpler to work with, where with yaml you have e.g. yaml hooks.

>You might justify it with "don't want to support," but even then that is resolved with the above statement "write a better SEO-optimized wiki, or add the problem to the issue template"

again, no one ever reads any of that, even with TOML all the docs are right there with examples and such, yet no one ever reads and asks ChatGPT or says on issues tracker how bad we are.

Also, keep in mind that unless someone rewrites _serde_(rust library for serialization) none of that is an easy task, because you have to convert between formats, and you can't convert yaml hooks at all, so you have to develop 2 internal reprs for all of that, which is just a madness and no-one has time for that thing (if you have go ahead and write serde alternative in rust for us, though by this point we'll sunset yaml config all together).

**Comment by anonhostpi** [comment]
The problem is that eventually, once Alacritty matures, this deprecated feature will turn right back into a feature request, since YAML is more widely adopted.

Also, "simple" is subjective. As you can see from the above, people will have complaints about TOML as well as YAML. People who are comfortable programming in Python and writing microservices (e.g. docker, gh-actions, etc) are going to find YAML simpler than TOML. People who deal with OS configuration files regularly are going to find TOML simpler than YAML. Using "simpler" or "better" when it comes to describing 2 languages that do the same things nearly identically is futile.

What you said about Rust's support for serialization languages does make sense, but then you end up with the problem in my first sentence. By the time that Alacritty matures, it is likely that Rust will have more reliable support for serialization languages.

But that's not even the main problem. It is very likely that people are going to request adding it back before Rust fixes serde, just simply because of it's wider adoption, end-user preference, and _established precedence_. As you said, people are going to _continue_ to ignore the docs, and just use copilot/gpt-generated files to install your software, which are all going to tell them to use YAML.

**Comment by chrisduerr** [comment]
> The problem is that eventually, once Alacritty matures, this deprecated feature will turn right back into a feature request

How is this a problem? There will always be a bunch of garbage feature requests, rejecting them is not an issue.

> What you said about Rust's support for serialization languages does make sense, but then you end up with the problem in my first sentence. By the time that Alacritty matures, it is likely that Rust will have more reliable support for serialization languages.

This just demonstrates lack of knowledge of the Rust ecosystem.

> So, instead of getting "YAML indentation" issue-spam, you're going to get "why is yaml deprecated?"/"FR: add YAML" issue-spam.

So instead of hunting bugs in user configs we can just close issues immediately. I don't see how this isn't an easy win.

**Comment by anonhostpi** [comment]
> > The problem is that eventually, once Alacritty matures, this deprecated feature will turn right back into a feature request
> 
> How is this a problem? There will always be a bunch of garbage feature requests, rejecting them is not an issue.

So is rejecting "indentation issues"

> > What you said about Rust's support for serialization languages does make sense, but then you end up with the problem in my first sentence. By the time that Alacritty matures, it is likely that Rust will have more reliable support for serialization languages.
> 
> This just demonstrates lack of knowledge of the Rust ecosystem.

Sure, but I have doubts that it is an inaccurate statement based off the lifetimes of other languages. Serialization is a basic language feature. I doubt that _any_ language community isn't going to work towards adopting de facto serialization standards.

> > So, instead of getting "YAML indentation" issue-spam, you're going to get "why is yaml deprecated?"/"FR: add YAML" issue-spam.
> 
> So instead of hunting bugs in user configs we can just close issues immediately. I don't see how this isn't an easy win.

You won't have to bug hunt yaml indentation problems, if you add yaml and toml linters to the issue template. If an end user doesn't lint their own files, then don't troubleshoot. It doesn't matter if its YAML or TOML.

**Comment by chrisduerr** [comment]
> From what I understand from the Rust ecosystem is that they do intend to offer competent YAML support: https://docs.rs/serde_yaml/latest/serde_yaml/

Version 0.8 of this library was using yaml-rs, which was abandoned because yaml is an awful file format that you cannot write parsers for. Version 0.9 switched to a Rust translation of libyaml, a C Library that was abandoned 3 years ago and has numerous known security issues.

So since you're clearly an expert in this field, please tell me more about the plethora of amazing deserialization options for yaml. It's such a cool format, I'd love to learn about more popular Rust libraries supporting it.

**Comment by anonhostpi** [comment]
> So since you're clearly an expert in this field, please tell me more about the plethora of amazing deserialization options for yaml. It's such a cool format, I'd love to learn about more popular Rust libraries supporting it.

I don't have to, because my earlier argument still stands:

> By the time that Alacritty matures, it is likely that Rust will have more reliable support for serialization languages.

**Comment by amne** [comment]
Docs say you can override shell with
```toml
[shell]
program="myshell"
args=["myargs","more"]
```

And then immediately after that it says `working_directory` is either a string or None but it's a root level field.
How do you parse this?
```toml
[shell]
program="myshell"
args=["myargs","more"]

working_directory="/home/whatever"
```

**Comment by kchibisov** [comment]
>How do you parse this?

The working directory ends up in the `[shell]` section, you should just put it above everything else.

**Comment by SidOfc** [comment]
Hello all, at first I liked my yaml config and I found that the generated toml version looked a bit clunky (not trying to dunk on tools which convert my config file for essentially free, thank you for saving me time! <3).

- [Initial alacritty.yml](https://github.com/SidOfc/dotfiles/blob/0eb9767af018b831c9d06fd23dfcd9d9ee85ae7b/alacritty.yml)
- [Generated alacritty.toml](https://github.com/SidOfc/dotfiles/blob/02e08258dba80388e8a2d0b729e9922608e75b1a/alacritty.toml)

So I went and found this issue because I was wondering about whether yaml support was actually being completely phased out (which seems to be desired by the OP). I figured the next best thing was to try and improve the readability of the toml file since that seems to be the way forward. I took a quick stroll through [the specs](https://toml.io/en/v1.0.0) which are actually remarkably easy to understand and came up with this version:

- [Current alacritty.toml](https://github.com/SidOfc/dotfiles/blob/45c3d3df7158c9fe78a5520d3952243cda6053fd/alacritty.toml)

Honestly it's quite good, I think I'm liking this more than my yaml config actually :)
Nobody particularly "loves" changes like this but OP has done the work to provide a deprecation warning accompanied by tooling to automatically get to a working toml version. Technically Alacritty is still in "beta" as well so it shouldn't come as a surprise that some changes will still be made.

All this is just to say please try to understand the OP's reasoning, take 5 minutes to improve the generated toml file using info from the specs and enjoy the result knowing it's very likely here to stay.

Finally, my apologies if this adds just more noise to the conversation. Just wanted to share some thoughts since there were some small arguments going on in the conversation.

**Comment by chrisduerr** [comment]
> All this is just to say please try to understand the OP's reasoning, take 5 minutes to improve the generated toml file using info from the specs and enjoy the result knowing it's very likely here to stay.

There's no point in manually writing toml formatters to try and get it into a subjectively better format when there's no way we'll ever match exactly the way it should be either. I used the automated tool to get started, then cleaned it up by hand, if people have an issue with that they can write their own migration tool.

**Comment by SidOfc** [comment]
That makes complete sense to me, I'm already happy about the fact that some simple conversion is included and just works. Taking a little bit of time to further refine automatically converted configuration it is a small price to pay for what is in my opinion the best terminal out there.

Thanks for all your work!

**Comment by amne** [comment]
> > How do you parse this?
> 
> The working directory ends up in the `[shell]` section, you should just put it above everything else.

of course. but how do you get out of section and back to root level? is toml limited here?

**Comment by chrisduerr** [comment]
> of course. but how do you get out of section and back to root level? is toml limited here?

Yeah this can be a bit confusing for people new to toml. Hopefully https://github.com/alacritty/alacritty/pull/7908 will resolve this but unfortunately it has stalled (for now). I'd like to see that land in Alacritty's next stable release.

**Comment by chrisduerr** [comment]
> Would you be willing to accept a PR that reinstates YAML as another option for specifying configuration, perhaps behind an "advanced" flag of some sort to reduce support burden that's alluded to in the issue?

No. If you don't like toml, use a different terminal. It will never be supported again.

**Comment by 0cjs** [comment]
While the switch can be justified, I suppose, it's a bit harsh to say that you cannot have a config file that works both on the current version of Debian (12) and on the latest version of alacritty. A compatibility period would have been nice.

**Comment by 0cjs** [comment]
Ah, ok. Living with a warning is justifiable, I think. @chrisduerr's comment above indicated that YAML is no longer accepted at all.

**Comment by chrisduerr** [comment]
> @chrisduerr's comment above indicated that YAML is no longer accepted at all.

That will happen in the future.
